File: src/utils/BPlusTree.ts
/**
 * B+ Tree Implementation in TypeScript
 * Ported from Python implementation provided by user.
 */

export class BPlusTreeNode<K, V> {
  keys: K[] = [];
  children: BPlusTreeNode<K, V>[] = [];
  values: V[] = [];
  next: BPlusTreeNode<K, V> | null = null;
  isLeaf: boolean;

  constructor(isLeaf = false) {
    this.isLeaf = isLeaf;
  }
}

export class BPlusTree<K, V> {
  t: number;
  maxKeys: number;
  minKeys: number;
  root: BPlusTreeNode<K, V>;

  constructor(order = 4) {
    this.t = order;
    this.maxKeys = 2 * order - 1;
    this.minKeys = order - 1;
    this.root = new BPlusTreeNode<K, V>(true);
  }

  search(key: K): V | null {
    let node = this.root;
    while (!node.isLeaf) {
      let i = 0;
      while (i < node.keys.length && key >= node.keys[i]) {
        i++;
      }
      node = node.children[i];
    }

    for (let i = 0; i < node.keys.length; i++) {
      if (node.keys[i] === key) {
        return node.values[i];
      }
    }
    return null;
  }

  insert(key: K, value: V): void {
    if (nodeIsFull(this.root, this.maxKeys)) {
      const oldRoot = this.root;
      const newRoot = new BPlusTreeNode<K, V>(false);
      newRoot.children.push(oldRoot);
      this.splitChild(newRoot, 0);
      this.root = newRoot;
    }
    this.insertNonFull(this.root, key, value);
  }

  private insertNonFull(node: BPlusTreeNode<K, V>, key: K, value: V): void {
    if (node.isLeaf) {
      for (let i = 0; i < node.keys.length; i++) {
        if (node.keys[i] === key) {
          node.values[i] = value;
          return;
        }
      }

      let i = node.keys.length - 1;
      node.keys.push(null as any);
      node.values.push(null as any);
      while (i >= 0 && node.keys[i] > key) {
        node.keys[i + 1] = node.keys[i];
        node.values[i + 1] = node.values[i];
        i--;
      }
      node.keys[i + 1] = key;
      node.values[i + 1] = value;
    } else {
      let i = node.keys.length - 1;
      while (i >= 0 && key < node.keys[i]) {
        i--;
      }
      i++;

      if (nodeIsFull(node.children[i], this.maxKeys)) {
        this.splitChild(node, i);
        if (key >= node.keys[i]) {
          i++;
        }
      }
      this.insertNonFull(node.children[i], key, value);
    }
  }

  private splitChild(parent: BPlusTreeNode<K, V>, index: number): void {
    const t = this.t;
    const child = parent.children[index];
    const newNode = new BPlusTreeNode<K, V>(child.isLeaf);
    const mid = t - 1;

    if (child.isLeaf) {
      newNode.keys = child.keys.slice(mid);
      newNode.values = child.values.slice(mid);
      child.keys = child.keys.slice(0, mid);
      child.values = child.values.slice(0, mid);

      newNode.next = child.next;
      child.next = newNode;

      parent.keys.splice(index, 0, newNode.keys[0]);
    } else {
      const midKey = child.keys[mid];

      newNode.keys = child.keys.slice(mid + 1);
      newNode.children = child.children.slice(mid + 1);
      child.keys = child.keys.slice(0, mid);
      child.children = child.children.slice(0, mid + 1);

      parent.keys.splice(index, 0, midKey);
    }
    parent.children.splice(index + 1, 0, newNode);
  }

  update(key: K, newValue: V): boolean {
    let node = this.root;
    while (!node.isLeaf) {
      let i = 0;
      while (i < node.keys.length && key >= node.keys[i]) {
        i++;
      }
      node = node.children[i];
    }

    for (let i = 0; i < node.keys.length; i++) {
      if (node.keys[i] === key) {
        node.values[i] = newValue;
        return true;
      }
    }
    return false;
  }

  getAll(): [K, V][] {
    const result: [K, V][] = [];
    let node = this.root;
    while (!node.isLeaf) {
      node = node.children[0];
    }

    let current: BPlusTreeNode<K, V> | null = node;
    while (current !== null) {
      for (let i = 0; i < current.keys.length; i++) {
        result.push([current.keys[i], current.values[i]]);
      }
      current = current.next;
    }
    return result;
  }

  clear(): void {
    this.root = new BPlusTreeNode<K, V>(true);
  }
}

function nodeIsFull<K, V>(node: BPlusTreeNode<K, V>, maxKeys: number): boolean {
  return node.keys.length === maxKeys;
}

File: Module_A/database/bplustree.py
import math

class Node:
    def __init__(self, is_leaf=False):
        self.is_leaf = is_leaf
        self.keys = []
        self.values = []
        self.children = []
        self.next = None

class BPlusTree:
    def __init__(self, order=4):
        self.root = Node(is_leaf=True)
        self.order = order

    def search(self, key):
        node = self.root
        while not node.is_leaf:
            i = 0
            while i < len(node.keys) and key >= node.keys[i]:
                i += 1
            node = node.children[i]
        
        for i, k in enumerate(node.keys):
            if k == key:
                return node.values[i]
        return None

    def insert(self, key, value):
        node = self.root
        if len(node.keys) == self.order - 1:
            new_root = Node()
            new_root.children.append(self.root)
            self._split_child(new_root, 0)
            self.root = new_root
        self._insert_non_full(self.root, key, value)

    def _split_child(self, parent, i):
        order = self.order
        node = parent.children[i]
        new_node = Node(is_leaf=node.is_leaf)
        mid = order // 2

        parent.keys.insert(i, node.keys[mid-1])
        parent.children.insert(i + 1, new_node)

        new_node.keys = node.keys[mid:]
        node.keys = node.keys[:mid]

        if node.is_leaf:
            new_node.values = node.values[mid:]
            node.values = node.values[:mid]
            new_node.next = node.next
            node.next = new_node
        else:
            new_node.children = node.children[mid:]
            node.children = node.children[:mid]

    def _insert_non_full(self, node, key, value):
        if node.is_leaf:
            i = 0
            while i < len(node.keys) and key > node.keys[i]:
                i += 1
            if i < len(node.keys) and node.keys[i] == key:
                node.values[i] = value
            else:
                node.keys.insert(i, key)
                node.values.insert(i, value)
        else:
            i = 0
            while i < len(node.keys) and key >= node.keys[i]:
                i += 1
            if len(node.children[i].keys) == self.order - 1:
                self._split_child(node, i)
                if key >= node.keys[i]:
                    i += 1
            self._insert_non_full(node.children[i], key, value)

    def update(self, key, value):
        self.insert(key, value)

    def delete(self, key):
        # Simplified delete: just remove from leaf
        node = self.root
        while not node.is_leaf:
            i = 0
            while i < len(node.keys) and key >= node.keys[i]:
                i += 1
            node = node.children[i]
        
        for i, k in enumerate(node.keys):
            if k == key:
                node.keys.pop(i)
                node.values.pop(i)
                return True
        return False

    def get_all(self):
        node = self.root
        while not node.is_leaf:
            node = node.children[0]
        
        results = []
        while node:
            for i in range(len(node.keys)):
                results.append({'key': node.keys[i], 'value': node.values[i]})
            node = node.next
        return results
